pkg/streamer/file_test.go

```diff
@@ -367,96 +378,83 @@ func (t *testUtilSuite) TestRelaySubDirUpdated(c *C) {
 	}()
 	wg.Wait()
 
-	// 3. new file created when adding watching
+	// 3. new binlog file created
 	err = ioutil.WriteFile(relayPaths[1], nil, 0o600)
 	c.Assert(err, IsNil)
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		relaySubDirUpdated(ctx, watcherInterval, subDir, relayPaths[0], relayFiles[0], 0, updatePathCh, errCh)
+		// update meta file
+		meta := Meta{BinLogName: relayFiles[1], BinLogPos: binlogPos, BinlogGTID: binlogGTID}
+		var buf bytes.Buffer
+		enc := toml.NewEncoder(&buf)
+		err2 := enc.Encode(meta)
+		c.Assert(err2, IsNil)
+		filename := path.Join(subDir, utils.MetaFilename)
+		err3 := utils.WriteFileAtomic(filename, buf.Bytes(), 0o644)
+		c.Assert(err3, IsNil)
+
+		relayLogUpdatedOrNewCreated(ctx, watcherInterval, subDir, relayPaths[0], relayFiles[0], 0, updatePathCh, errCh)
 		c.Assert(len(errCh), Equals, 0)
 		c.Assert(len(updatePathCh), Equals, 1)
 		up := <-updatePathCh
 		c.Assert(up, Equals, relayPaths[1])
 	}()
 	wg.Wait()
 
-	// 4. file updated
+	// 4. file increased when checking meta and meta file updated
 	wg.Add(1)
 	go func() {
 		defer wg.Done()
-		relaySubDirUpdated(ctx, watcherInterval, subDir, relayPaths[1], relayFiles[1], 0, updatePathCh, errCh)
+		// update meta file to new file
+		meta := Meta{BinLogName: relayFiles[2], BinLogPos: binlogPos, BinlogGTID: binlogGTID}
+		var buf bytes.Buffer
+		enc := toml.NewEncoder(&buf)
+		err2 := enc.Encode(meta)
+		c.Assert(err2, IsNil)
+		filename := path.Join(subDir, utils.MetaFilename)
+		err3 := utils.WriteFileAtomic(filename, buf.Bytes(), 0o644)
+		c.Assert(err3, IsNil)
+		fi, err4 := os.Stat(relayPaths[1])
+		c.Assert(err4, IsNil)
+		currSize := fi.Size()
+		c.Assert(failpoint.Enable("github.com/pingcap/dm/pkg/streamer/CMPAlwaysReturn0", `return(true)`), IsNil)
+		c.Assert(failpoint.Enable("github.com/pingcap/dm/pkg/streamer/DoNotReturnEvenMetaChange", `return(true)`), IsNil)
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			relayLogUpdatedOrNewCreated(ctx, watcherInterval, subDir, relayPaths[1], relayFiles[1], currSize, updatePathCh, errCh)
+		}()
+		// write old file
+		err5 := ioutil.WriteFile(relayPaths[1], data, 0o600)
+		c.Assert(err5, IsNil)
 		c.Assert(len(errCh), Equals, 0)
-		c.Assert(len(updatePathCh), Equals, 1)
 		up := <-updatePathCh
 		c.Assert(up, Equals, relayPaths[1])
 	}()
-
-	// wait watcher started, update the file
-	time.Sleep(2 * watcherInterval)
-	err = ioutil.WriteFile(relayPaths[1], data, 0o600)
-	c.Assert(err, IsNil)
-	wg.Wait()
-
-	// 5. file created
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		relaySubDirUpdated(ctx, watcherInterval, subDir, relayPaths[1], relayFiles[1], size, updatePathCh, errCh)
-		c.Assert(len(errCh), Equals, 0)
-		c.Assert(len(updatePathCh), Equals, 1)
-		up := <-updatePathCh
-		c.Assert(up, Equals, relayPaths[2])
-	}()
-
-	// wait watcher started, create a new file
-	time.Sleep(2 * watcherInterval)
-	err = ioutil.WriteFile(relayPaths[2], data, 0o600)
-	c.Assert(err, IsNil)
 	wg.Wait()
 
-	// 6.
```

Comment:
why remove these tests?

Reply:
Because the logic of this function has been changed

The test section for file created on line 409 is already above (line 389)

The logic for listening for directory-related events on line 426 is also gone